include/swift/SIL/SILInstruction.h

```diff
@@ -3266,6 +3268,13 @@ class BeginAccessInst
     SILInstruction::Bits.BeginAccessInst.Enforcement = unsigned(enforcement);
   }
 
+  bool isTracking() const {
+    return SILInstruction::Bits.BeginAccessInst.Tracking;
+  }
+  void setTracking(bool tracking) {
+    SILInstruction::Bits.BeginAccessInst.Tracking = tracking;
+  }
```

Comment:
This seems like the right canonical place to document what this means and how it changes the invariants of SIL.  (I assume it does not change the invariants of SIL and that there still needs to be an end_access; it just doesn't translate to anything at runtime anymore.)

Is "tracking" a good name for this?  It feels like a description of the wrong level of abstraction.  The meaning of the flag is whether it's possible for conflicting accesses to the variable to synchronously happen within the duration of the access.

So you definitely think that tracking this in SIL is the right thing to do vs. just doing a final analysis in IRGen?  Or is the plan to do that final analysis in IRGen, but just as a separate pass rather than literally doing it in IRGenSIL.cpp?

Reply:
The tracking flag doesn't change SIL structural properties. No SIL passes/utilities need to know about this flag unless they want to optimize it. The slight exception is that if we mark begin_unpaired_access as [nontracking] then we need to delete the end_unpaired_access.

That's also why I didn't make it an enforcement level. It's orthogonal to the rest of the code that operates on access markers. Although I admit it's not a useful flag unless the enforcement ends up being dynamic.

> Is "tracking" a good name for this? It feels like a description of the wrong level of abstraction. The meaning of the flag is whether it's possible for conflicting accesses to the variable to synchronously happen within the duration of the access.

I think it's a good name at the runtime interface, since it will be used for more than just those semantics. It literally means nothing more than: this access will not be dynamically checked against any subsequent access, and the end_access is invisible to the runtime.

I never considered that [nontracking] would imply those no-synchronous-access semantics, so much as those semantics would imply that it's safe to mark the instruction as [nontracking].

So, I would say that the current patch is really exposing a runtime feature rather than declaring semantic properties.

The alternative name would be something like "instantaneous". But I didn't like that notion considering we still have begin/end scopes. If you can think of a better name for "enforced until end of scope" let me know. But I thought it best to use the same name for the runtime interface and SIL flag.

> So you definitely think that tracking this in SIL is the right thing to do vs. just doing a final analysis in IRGen? Or is the plan to do that final analysis in IRGen, but just as a separate pass rather than literally doing it in IRGenSIL.cpp?

I wouldn't have gone to the trouble of changing SIL otherwise. I don't want to limit the optimization to whatever is easy enough to do in IRGenSIL.cpp (in addition to being generally opposed to adding nontrivial optimization to IRGen). I suspect that this will benefit from the same interprocedural analysis that will also give us better dynamic-to-static enforcement selection.

Also, I don't think having IRGen information helps here unless I'm missing something. It's just a SIL optimization that can run as late as we want, because nothing else in SIL depends on it.

Reply:
I agree that IRGen information doesn't help here.  My only concern about running it earlier in the pipeline is that operations might be moved into the access and invalidate the analysis.  I guess that's prohibited anyway if it might introduce new conflicts.

As for the name, maybe something about how no conflicting accesses can start during the access?  Something about "overlap"?  "disjoint"?

Reply:
Ok. I suppose if this is going to be a SIL attribute (which it should be), then it's useful to be able to infer semantics from it. i.e. if the flag is set, SIL passes can assume no nested conflicts without rerunning the analysis.

Terms like disjoint/subaccess are already used for projections. We want to convey that there is no inner scope conflict (there may be an outer scope conflict).

Can I just use [safescope] instead of [nontracking]? Then explain the meaning in the doc comments?

You're ok with ExclusivityFlags::Tracking for the runtime interface right? This will be used to disable tracking when the end access may move threads.

Reply:
`[safescope]` sounds like it ought to be a military intelligence operation. :)

We could just use `[no_nested_conflicts]`.  Wordy but precise.

Reply:
I totally agree.